### PR TITLE
Fix LRU cache problem

### LINKED ISSUE #1039: LRUCache is not thread safe
The `LRUCacheStore` default implementation extends from  `LinkedHashMap`, which is not a thread safe collection. It's a problem since even the `get` operation is considered a **structural modification**, as described in his [javadoc](https://docs.oracle.com/javase/7/docs/api/java/util/LinkedHashMap.html) and can cause locks when various threads even try to read from it.

Also, the `LRUCacheFactory` returns `LRUCacheStore` from the producer method instead of the `CacheStore` interface. In that way, it's not possible to override the default implementation using a simple `@Alternative` (I don't know why, but CDI refuse to use my implementation instead of the default).

Sugested fix:

```java
public Cache getCache(InjectionPoint ip){
  int capacity = ip.getAnnotated().getAnnotation(LRU.class).capacity();
    Cache<Object, Object> guavaCache = CacheBuilder.newBuilder()
	.concurrencyLevel(50) //number of threads, may be configurable, may be omitted
	.maximumSize(capacity)
	.build();
	return new GuavaCacheWrapper(guavaCache); //just a delegate/wrapper
}
```

### PR DESCRIPTION
**Pull request by mariofts**
Solves the problem described in #1039.
## Changelog: 

- Creates a new implementation of `CacheStore` using a wrapper to the Guava's `Cache`. As Guava's `Cache` is tread safe by default, this resolves #1039.

### CODE REVIEW

File: vraptor-core/src/main/java/br/com/caelum/vraptor/cache/CacheStoreFactory.java [hunk 1]
```diff
@@ -31,14 +41,58 @@
...
 	@Default
-	public <K, V> DefaultCacheStore<K, V> buildDefaultCache() {
+	public <K, V> CacheStore<K, V> buildDefaultCache() {
 		return new DefaultCacheStore<K, V>();
 	}
 
 	@Produces
 	@LRU
-	public <K, V> LRUCacheStore<K, V> buildLRUCache(InjectionPoint ip) {
+	public <K, V> CacheStore<K, V> buildLRUCache(InjectionPoint ip) {
```
Comment (Turini) [inline vraptor-core/src/main/java/br/com/caelum/vraptor/cache/CacheStoreFactory.java]: nice catch!

File: vraptor-core/src/main/java/br/com/caelum/vraptor/cache/CacheStoreFactory.java [hunk 2]
```diff
@@ -31,14 +41,58 @@
...
+				return guavaCache.get(key, new Callable<V>() {
+
+					@Override
+					public V call() throws Exception {
+						return userSuplier.get();
+					}
+				});
+			} catch (ExecutionException e) {
+				propagateIfPossible(e);
+				return valueProvider.get();
```
Comment (Turini) [inline vraptor-core/src/main/java/br/com/caelum/vraptor/cache/CacheStoreFactory.java]: @mariofts we don't need to throw CacheException here? I didn't get this change.

**Comment by Turini** [comment]
@mariofts your PR sounds amazing, thx for doing this. You can :sheep: after checking my comments :)

File: vraptor-core/src/main/java/br/com/caelum/vraptor/cache/CacheStoreFactory.java (hunk 2 above)
Comment (mariofts) [inline vraptor-core/src/main/java/br/com/caelum/vraptor/cache/CacheStoreFactory.java]: done

**Comment by Turini** [comment]
thx @mariofts!

**Pull request merged**